From 6ff9d26e7a21b198323874a29c7866d557f16746 Mon Sep 17 00:00:00 2001
From: Admin <lockedthreaddev@gmail.com>
Date: Sun, 5 May 2019 19:32:33 -0400
Subject: [PATCH] Fixed buckets breaking all blocks


diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index 0099598de..49fc804d2 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -239,7 +239,7 @@ public class BlockFlowing extends BlockFluids {
         Block block = world.getType(blockposition).getBlock();
 
         if (!LockedSpigotConfig.isLiquidsBreakRedstone()) {
-            return block instanceof BlockRedstoneWire || block instanceof BlockRedstoneTorch || block instanceof BlockRedstoneComparator;
+            return block instanceof BlockRedstoneWire || block instanceof BlockRedstoneTorch || block instanceof BlockRedstoneComparator || block instanceof BlockDoor || block == Blocks.STANDING_SIGN || block == Blocks.LADDER || block == Blocks.REEDS || (block.material == Material.PORTAL || block.material.isSolid());
         }
 
         return block instanceof BlockDoor || block == Blocks.STANDING_SIGN || block == Blocks.LADDER || block == Blocks.REEDS || (block.material == Material.PORTAL || block.material.isSolid());
-- 
2.21.0.windows.1

