From ab73867d87f7312b40073963bb5294e9b278bdc9 Mon Sep 17 00:00:00 2001
From: Admin <lockedthreaddev@gmail.com>
Date: Sat, 11 May 2019 23:03:47 -0400
Subject: [PATCH] Fixed test command


diff --git a/src/main/java/dev/lockedthread/commands/CommandTest.java b/src/main/java/dev/lockedthread/commands/CommandTest.java
index 7b0a15588..37742fc8b 100644
--- a/src/main/java/dev/lockedthread/commands/CommandTest.java
+++ b/src/main/java/dev/lockedthread/commands/CommandTest.java
@@ -38,7 +38,7 @@ public class CommandTest extends Command {
                         } else {
                             sender.sendMessage(ChatColor.RED + "You must be a player to do this!");
                         }
-                        break;
+                        return true;
                     case "tnt":
                         if (sender instanceof Player) {
                             int amount = args.length == 2 ? Integer.parseInt(args[1]) : 100;
@@ -49,9 +49,10 @@ public class CommandTest extends Command {
                         } else {
                             sender.sendMessage(ChatColor.RED + "You must be a player to do this!");
                         }
+                        return true;
                     default:
                         sender.sendMessage(ChatColor.RED + "Unable to find EntityType with name \"" + args[0] + "\"");
-                        break;
+                        return true;
                 }
             }
         }
-- 
2.21.0.windows.1

